### PR TITLE
Feature/randfft cherry pick

### PR DESCRIPTION
**Pull request by bjoo**
Cherry picked the cuFFT, cuRand and quda_fp16 refactors from my hip oriented WIP branch.
Code built on Summit, but some tests on Summit had errors... @maddyscientist said she hasn't checked it for correctness recently. I'll roll back to her version (from before my refactor) and check if that had an issue, or whether what I picked up is a Summit issue.

### CODE REVIEW

File: include/kernels/pgauge_init.cuh [hunk 1]
```diff
@@ -121,7 +121,11 @@ namespace quda {
     aabs = sqrt( 1.0 - a(0,0) * a(0,0));
     ctheta = Random<T>(localState, (T)-1.0, (T)1.0);
     phi = PII * Random<T>(localState);
-    stheta = ( curand(&localState) & 1 ? 1 : -1 ) * sqrt( (T)1.0 - ctheta * ctheta );
+
+    // Was   xurand(*localState>& 1 ? 1 : -1
+    // which presumably just selects when the lowest bit is 1 or 0 with 50% probability each
+    // so this should do the same, without an appeal to the bit swizzle, but may end up being slower.
+    stheta = ( uniform<T>::rand(localState) < static_cast<T>(0.5) ? 1 : -1 ) * sqrt( (T)1.0 - ctheta * ctheta );
```
Comment (maddyscientist) [inline include/kernels/pgauge_init.cuh:128]: What's the motivation for this change?

File: include/targets/hip/random_quda.h [hunk 2]
```diff
@@ -0,0 +1,152 @@
+#pragma once
```
Comment (maddyscientist) [inline include/targets/hip/random_quda.h]: There's a lot of repeated boilerplate in here and the `CUDA` equivalent.  While of course the `hiprandState` vs `cuRNGState` needs to be in the specialization, why not move the class `RNG` to the generic header?  We could make the class templated, with the type used to instantiate the class either `cuRNGState` or `hiprandState` in the respective target specialization?

**Comment by bjoo** [comment]
That is possible to do, but I’d rather do it in a cleanup as I am keen to return to the other branch
I am working on.

Best,
B

--
Balint Joo,  Oak Ridge Leadership Computing Facility,
Oak Ridge National Laboratory
P.O. Box 2008, 1 Bethel Valley Road, Oak Ridge, TN 37831, USA
email: joob AT ornl.gov.  Tel: +1-757-912-0566 (cell, remote)

**Comment by bjoo** [comment]
We discussed this in the meeting.
pgauge_init originally called curand() directly (which was API specific) and the
purpose appeared to be purely to make a 50-50 choice (based on the lowest bit
being 1 or 0). This change

  *   removes the API specificity by calling the API independent uniform() function
  *   makes it clear that we are trying to make a 50-50 choice by trying to “throw under” 0.5 rather than bitmasking
  *

--
Balint Joo,  Oak Ridge Leadership Computing Facility,
Oak Ridge National Laboratory
P.O. Box 2008, 1 Bethel Valley Road, Oak Ridge, TN 37831, USA
email: joob AT ornl.gov.  Tel: +1-757-912-0566 (cell, remote)

**Comment by bjoo** [comment]
As discussed on Slack:
* moved RNG class back into `include/random_quda.h` since it is common between QUDA and HIP targets (may end up being completely arch independent)

* as result renamed `include/targets/cuda/random_quda.h` and `include/targets/hip/random_quda.h` to be `random_quda_impl.h` respectively in both directories

* got rid of `Random<T>(state)` and `Random<T>(state,a,b)`  since these are uniform random deviates implemented identically in `uniform<T>::rand(state)` (and now also `uniform<T>::rand(state,a,b)`. What is more with care similar functionality (to scale and translate a distribution could be implemented in `normal<T>::`. This reduced the size of the CUDA/HIP specific implementations in the `random_quda_impl.h` files.
* trawled the code and eliminated calls to `Random<T>(...)` and replace them with `uniform<T>::rand(...)`
* Code compiled on summit and everything passed except 
```The following tests FAILED:
	  2 - blas_test_full_wilson (Failed)
```
which also failed for me on summit in the regular `feature/generic_kernel` branch prior to merges. Filed this as issue #1070

**Comment by bjoo** [comment]
With latest merge commit, all tests passed on Summit! I hope this should be good to go now.
